src/samples/audio/smart_amp_llext/CMakeLists.txt

```diff
@@ -0,0 +1,71 @@
+# Copyright (c) 2023 Intel Corporation.
+# SPDX-License-Identifier: Apache-2.0
+
+cmake_minimum_required(VERSION 3.20.0)
+find_package(Zephyr REQUIRED HINTS $ENV{ZEPHYR_BASE})
+project(smart_amp_test)
+
+SET_PROPERTY(GLOBAL PROPERTY TARGET_SUPPORTS_SHARED_LIBS TRUE)
+
+set(MODULE "smart_amp_test")
+cmake_path(SET SOF_BASE NORMALIZE ${PROJECT_SOURCE_DIR}/../../../..)
+
+add_library(${MODULE} SHARED)
+
+target_sources(${MODULE} PRIVATE
+	${CMAKE_CURRENT_LIST_DIR}/../smart_amp_test_ipc4.c
+)
+
+sof_append_relative_path_definitions(${MODULE})
+
+target_include_directories(${MODULE} PRIVATE
+	"${ZEPHYR_BASE}/include"
+	"${ZEPHYR_BASE}/soc/xtensa/intel_adsp/common/include"
+	"${ZEPHYR_BASE}/soc/xtensa/intel_adsp/ace/include/intel_ace15_mtpm"
+	"${ZEPHYR_BASE}/../modules/hal/xtensa/include"
+	"${ZEPHYR_BASE}/../modules/hal/xtensa/zephyr/soc/intel_ace15_mtpm"
+	"${SOF_BASE}/src/include"
+	"${SOF_BASE}/src/arch/xtensa/include"
+	"${SOF_BASE}/src/platform/meteorlake/include"
+	"${SOF_BASE}/src/platform/intel/ace/include"
+	"${SOF_BASE}/src/include/sof/audio/module_adapter/iadk"
+	"${SOF_BASE}/zephyr/include"
+	"${SOF_BASE}/xtos/include"
+	"${SOF_BASE}/tools/rimage/src/include"
+	"${PROJECT_BINARY_DIR}/../include/generated"
+)
+
+set(MODULE_PROPERTIES HPSRAM_ADDR "0xa06c1000")
+set_target_properties(${MODULE} PROPERTIES ${MODULE_PROPERTIES})
+
+set(MODULE_COMPILE_DEF
+	__ZEPHYR__=1
+	__XTENSA__
+	KERNEL
+	MAJOR_IADSP_API_VERSION=5
+	MIDDLE_IADSP_API_VERSION=0
+	MINOR_IADSP_API_VERSION=0
+)
+target_compile_definitions(${MODULE} PRIVATE ${MODULE_COMPILE_DEF})
+
+target_compile_options(${MODULE} PRIVATE
+	-imacros${PROJECT_BINARY_DIR}/../include/generated/autoconf.h
+	-save-temps -O2
+)
+
+set(MODULE_LINKER_PARAMS -nostdlib -nodefaultlibs)
+target_link_options(${MODULE} PRIVATE
+	${MODULE_LINKER_PARAMS}
+)
+
+add_custom_target(${MODULE}_llext ${CMAKE_C_COMPILER}
```

Comment:
> Every `add_custom_target()` needs a corresponding `add_custom_command() and vice-versa, this is tedious, painful and error-prone but this CMake for you: https://samthursfield.wordpress.com/2015/11/21/cmake-dependencies-between-targets-and-files-and-custom-commands/
> 
> ``

@marc-hb sorry, not getting it why "every `add_custom_target()` needs a corresponding `add_custom_command()`?" `add_custom_target()` can include commands to build the target already, AFAIU `add_custom_command()` is only needed if one needs to *add more* commands (to that target)?

Reply:
The best long story is https://samthursfield.wordpress.com/2015/11/21/cmake-dependencies-between-targets-and-files-and-custom-commands/

The short story is: because CMake sucks.

Reply:
> @teburd, @tejlmand is every llext module expected to implement this usual but ugly add_custom_ dance by itself?

initial look tells me it should be possible to create a common function which defines the needed target, but would need to know a bit more on the exact use-case for this to be able to provide some guidelines on doing so, if there is a need.

Reply:
> Every add_custom_target() needs a corresponding `add_custom_command() and vice-versa.

Not true, example from Zephyr, `menuconfig` is a custom target not needed anything else.
But when multiple other targets depends on output created by a single custom command, then such command must be driven by a custom target. Something which often true in complex build system, like Zephyr.

Reply:
> The best long story is https://samthursfield.wordpress.com/2015/11/21/cmake-dependencies-between-targets-and-files-and-custom-commands/
> 
> The short story is: because CMake sucks.

@marc-hb sorry, the short story is too short and the long one is too long. I looked through the whole article (sorry again, didn't take the time to read it completely), and didn't find the explanation to why this is needed. Could you point out to a specific section in it, please?

Reply:
I am fine with a custom rule today. The important thing is that we have modules and can re-visit cmake improvements when they become priority.

Reply:
> Could you point out to a specific section in it, please?

Not really sorry. If this could be made shorter then:
- it wouldn't have taken me years of CMake practice to finally realize why you always must use both `add_custom_command()` + `add_custom_target()` at the same time
- I wouldn't keep forgetting the detailed reasons why and I wouldn't need a bookmark to this 1700 words long article.

Here's 2 sentences from the introduction:

> If you’ve tried do anything non-trivial in CMake using the add_custom_command() command, you may have got stuck in this horrible swamp of confusion [...] What makes it so hard is that there’s not one limitation, but several. 


The good news is you don't actually need to read and understand all the issues in this article, instead you can just trust me and this article and just use both `add_custom_command()` + `add_custom_target()`. That's exactly what I do myself; as I just wrote I don't remember precisely all the bad things that happen when you use only one. We're getting close to the point where doing it would take less time than discussing it.

The very slightly less short story is: **dependencies are not expressed correctly** if you don't use both which means things are not always built when you expect them to be (below the surface, CMake is a way to express dependencies like any other build system). So either you run out of date code or you're faced with a super cryptic error message, buried by the build concurrency. Either wastes an incredible amount of time.

Reply:
> The very slightly less short story is: **dependencies are not expressed correctly** if you don't use both which means things are not always built when you expect them to be

@marc-hb aha, ok, that's already something, not an explanation but at least a hint at what might happen otherwise... Since this is supposedly a bug in cmake (i.e. it isn't a documented feature, it's an empirically derived work-around), this might also behave differently with different versions of cmake, and hopefully some newer ones might even actually fix it, or it might not trigger in this specific case at all, or it might be something else, but since it doesn't cost much, we can add it here.

Reply:
> Since this is supposedly a bug in cmake (i.e. it isn't a documented feature, it's an empirically derived work-around), this might also behave differently with different versions of cmake,

No, this is a _design_ issue with respect to custom rules. Different CMake versions should behave mostly the same. What does not behave the same is... users: different users have this weird tendency to modify different files! To which a half-broken graph of dependencies will react differently. Then, users will ALL be consistently confused no matter what happened and they will reach the "CMake hotline" :-)